<a href="https://colab.research.google.com/github/angegonzalez/DiscreteMath/blob/master/RSA_MDII.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# RSA (Rivest, Shamir y Adleman)

Universidad Nacional de Colombia 
<br/>
2020-I | **Matemáticas** **Discretas**
<br/>
*Por:* Angel Mateo González Bejarano




El **RSA** es un sistema criptográfico de llave pública desarrollado en el año de 1979 por Rivest, Shamir y Adleman en MIT. Es uno de los algoritmos de este tipo más utilizados. En un sistema de criptografía de llave publica cada usuario posee dos llaves, una pública y otra privada. 
Cuando se quiere enviar un mensaje, el emisor busca la clave pública del receptor, cifra su mensaje con esa clave, y una vez que el mensaje cifrado llega al receptor, este se ocupa de descifrarlo usando su clave privada. La seguridad del algoritmo se basa en la dificultad de resolver el problema de **factorización de números enteros**.



**Preguntas planteadas**

1. **Problema.** Se debe especificar cuál es el problema que resuelve RSA, y por qué es necesario introducir un algoritmo de este tipo.

El algoritmo RSA es uno de los más utilizados en la actualidad en cuanto a temas de ciberseguridad, ya que permite que en un entorno en el cual exista una comunicacion de dos personas (que en principio se asume es inseguro, es decir, un entorno en el cual sea posible interceptar la comunicacion) esta misma se pueda dar de una forma segura, encriptando el mensaje de modo que si es interceptable este no pueda ser de alguna forma descifrado. 
<br/>
El ejemplo clásico para explicar este algoritmo es el de Alice y Bob, dos personas que se quieren comunicar, pero se quiere que la transmision del mensaje, por ejemplo, de Bob a Alice se haga forma segura y esto quiere decir que solo Alice pueda saber el mensaje que le está enviando Bob.
<br/>
RSA plantea un sistema de clave publica y privada, con esto, Bob encripta el mensaje usando informacion de la clave pública de Alice y luego ella puede desencriptar el mensaje usando su clave privada.
<br/>
Más adelante se explicara en detalle el funcionamiento de este algoritmo, pero la idea central es generar cierta seguridad sobre el mensaje, y esta seguridad esta dada por la **factorización de numeros enteros**, ya que este algoritmo trabaja con el producto de dos numeros primos (generalmente dados del orden de $10^{300}$) escogidos de manera aleatoria, y dado un resultado muy grande de este producto es virtualmente imposible conocer la factorización de este.
<br/>
Actualmente, el poder de computo esta avanzando a grandes pasos y hemos llegado al punto de inventar nuevas computadoras, como las cúanticas, que plantean una mejora significativa en los procesos de computo y operaciones. Con esto, valdría la pena preguntarse si la seguridad en la que se basa este algoritmo, puede llegar a quedar obsoleta en el escenario en el que este tipo de computadoras logren hacer los cálculos de una manera mas rapida y eficiente, es decir, si se puede llegar a resolver el problema de la factorizacion de enteros, este sistema quedaría obsoleto. 

2. **Desarrollo matemático del algoritmo.** Se deben especificar todos los pasos del algoritmo y se debe explicar de forma detallada cada uno de los pasos. Si el algoritmo hace uso de algunos resultados formales (teoremas) estos deben enunciarse y probarse de ser necesario.

Volvamos a nuestro ejemplo en el que Bob queire enviarle un mensaje a Alice, entonces definimos los siguientes pasos: 

#### **Previo al cifrado**
1. Alice escoge dos primos $p$ y $q$ (de más de 100 digitos) y calcula: $n=pq$ y $\phi(n) = (p-1)(q-1) $
  <br>
  Para este paso, el algoritmo debe seleccionar $p$ y $q$ aleatoriamente, para esto usamos el llamado test de primalidad que se basa en el postulado de Bertrand, el cual dice que: <br>

  > Si $n$ es un entero y $n>1$, entonces existirá al menos un numero primo $p$ tal que  $n< p < 2n $
  
  En el codigo veriamos lo siguiente (esto es usado a partir del ejemplo 2):



In [0]:
import random 
import math

def is_prime(x): #Verifica si un numero es primo o no
    if (x<=1):
        return False
    for i in range(2, math.ceil(math.sqrt(x))+1):
        if(x % i==0 and i!= x):
            return False
    return True
def get_prime(x): #Test de primalidad (Genera primos aleatorios)
  p = random.randint(x,2*x)
  while( not is_prime(p) ):
    p+=1
    if(p == 2*x):
      p = n 
  return p

p= get_prime(1000000000)
q= get_prime(1073200000)
print('Ejemplo de dos numeros primos aleatorios de 10 cifras: {}, {}'.format(str(p),str(q)))

Ejemplo de dos numeros primos aleatorios de 10 cifras: 1509069761, 1080855059



> Para el calculo de $\phi(n)$ tomamos en cuenta dos propiedades de la funcion: 

> 1. Si $n$ y $m$ son primos entre sí, entonces $\phi(nm)= \phi(n) \phi(m)$ <br/>
   **Demostración:**<br/>
   $\phi(nm) = |\{ a \in \mathbb{Z^+} \ ; \ a<nm \ \wedge mcd(nm, a)= 1 \}|$ <br/>
   $mcd(nm, a) = 1 \leftrightarrow mcd(n,a) = 1 \ \wedge \ mcd(m,a) =1 $ <br/>
   $\phi(nm) = |\{ a \in \mathbb{Z^+} \ ; \ a<nm \ \wedge mcd(n, a)= 1 \ \wedge \ mcd(m,a) =1 \}$
   <br/>
   Se pueden organizar los elementos de tal manera que:
   <br/>
   <br/>
   <img src="https://drive.google.com/uc?id=1HB6wjqeDiBH9HVkb8AQAdLARbth1bxwf"/>
   <br/>
   <br/>
   Los anteriores numeros son de la forma $qm+r$ para cierto $q,r \ \in \mathbb{Z}$
   <br/>
   Decimos entonces que se tiene $$ mcd(qm+r, m)= mcd(r,m)$$ <br/>
   Luego, nos preguntamos cuantas columnas hay con primos entre sí con $m$ y esto en efecto es $\phi(m)$. <br/>
   También, nos preguntamos dentro de una columna cuantos numeros son primos entre sí con $n$, en efecto habrán $\phi(n)$ <br/>
   Por lo tanto queda demostrado que $\ \phi(nm) =\phi(n) \phi(m) $ <br/>
2. Por definicion si $n$ es primo, entonces $\phi (n) = n-1$



2. Alice escoge una clave de cifrado $e$ tal que $$ mcd(e, \phi(n))=1$$

3. Alice calcula $d$ (como parte de su clave privada) con la propiedad que: $$ed \equiv 1 \ mod (\phi(n)) $$

4. La clave publica de Alice será $(e,n)$ y la clave privada, la cual no compartirá será $(d,n)$

#### Cifrado
1. Bob convierte el mensaje $M$ en un numero entero $P$ menor que $n$ mediante un protocolo acordado garantizando siempre que $P$ y $n$ sean primos entre sí

2. Bob aplica la transformación $$ C \equiv P^e (\ mod \ n\ )$$

3. Bob envia $C$ a Alice

#### Descifrado
1. Alice recibe $C$ de parte de Bob, y calcula: $$M \equiv C^d (\ mod \ n\ )$$

  Lo anterior funciona ya que $C^d \equiv (P^e)^d = P^{ed} = P^{k\phi(n)+1} = P(P^{\phi(n)})^k \equiv P (\ mod \ n\ )$. <br/>
  Ya que el según Teorema de Euler: $P^{\phi(n)} \equiv 1 (\ mod \ n\ )$ cuando $mcd (P,n) =1$ <br/><br/>
  **Teorema de Euler**
> Si $a$ y $n$ son enteros primos entre sí, entonces $a^{\phi(n)} \equiv 1 (\ mod \ n\ )$

  **Demostración:** <br/>
  Sea $P = \{ x \in \mathbb{Z^+}; \ x<n \ \wedge \ mcd(x,n)=1 \}$ <br/>
  Sea $Q = \{ y \in \mathbb{Z^+}, x \in P; \  y = a \cdot x \}$ <br/><br/>
  Los elementos del conjunto $Q$ son congruentes a los del conjunto $P$ en diferente orden.<br/>
  Luego, sea $u$ el producto de los elementos de $P$ y sea $v$ el producto de los elementos de $Q$.<br/>
  Los numeros $v$ y $u$ son congruentes dado que sus factores lo son: $$v \equiv u (\ mod \ n\ )$$<br/>
  Luego se tiene que $v = u \cdot a^{\phi(n)}$ <br/>
  Asi cancelamos el factor $u$ en la congruencia $$v \equiv u (\ mod \ n\ ) \rightarrow u \cdot a^{\phi(n)}  \equiv u (\ mod \ n\ ) $$<br/>
  Al final concluimos que $$a^{\phi(n)} \equiv 1 (\ mod \ n\ )$$


## **Ejemplo del algoritmo**
3. Se debe incluir un ejemplo numérico sencillo que aclare el funcionamiento del algoritmo.

  Vamos a retomar los pasos descritos anteriormente, para ello Bob quiere enviarle un mensaje a Alice, en principio este ejemplo se hace con valores numéricos sencillos con el ánimo de explicar el funcionamiento y que los cálculos no se tornen tediosos.


  >> **Previo al cifrado**

  >> Definimos $p$ y $q$ (numeros primos) como : <br/>
    $p= 47$ , $q=61$

  >> Definimos $n$ como el producto entre $p$ y $q$, en efecto: <br/>
    $n=2867$ <br/>
    También calculamos $\phi(n)$: <br/>
    $\phi(n)= 2760 $

  >> Luego, elegimos $e$ tal que $mcd(e,\phi(n)) =1 $ <br/>
    En este caso, escogemos $e$= 7 

  >> Calculamos $d$ con la propiedad $ed \equiv 1 \ mod (\phi(n)) $: <br/>
    Para ello, nos vamos a apoyar de un pequeño codigo que verifica esta condicion y nos da el resultado de $d$:


In [0]:
#Obtenemos d, para el ejemplo e=7, phi_n= 2760
def get_d_key(e, phi_n ):
     c = 1
     while (c % e > 0):
      c += phi_n
     return c // e
print("La clave de desciframiento d es: {}".format(str(get_d_key(7, 2760)))) 


La clave de desciframiento d es: 1183



>> Obtenemos así $d= 1183$

>> Así la clave publica de Alice es $(7, 2867)$ la privada será $(1183, 2867)$<br/>

> Para el siguiente paso, necesitamos convertir el mensaje $M$ a un equivalente numérico para eso vamos a seguir el siguiente esquema

<img src="https://drive.google.com/uc?id=1mCiLouHSxAZXPfjqMWddJ9gj2b4fSAP0"/>

>Para este ejemplo nuestro mensaje será: *UNAL* <br/>
Aplicando la transformacion tenemos el mensaje numérico: 

        2113 0011

>> **Cifrado** <br/>
  Aplicamos la transformacion $ C \equiv P^e (\ mod \ n\ )$ <br/>
  Luego de hacer los cálculos vemos que el mensaje cifrado es: 

            0859 0172
        
>> Finalmente, enviamos el mensaje.



>> **Descifrado** <br/>
>> Calculamos $M \equiv C^d (\ mod \ n\ )$ <br/> 
  Después de hacer los cálculos correspondientes obtenemos:

            2113 0011
>> Que corresponde al mensaje que Bob había enviado.


## **Implementación del algoritmo**
El algoritmo debe implementarse en lenguaje pyhton (sin excepción). La sección de implementación debe incluir un conjunto de ejemplos del uso del algoritmo.



---
A continuacion propongo una implementacion del algoritmo RSA en `python`:

---




La clase `Receptor` permite inicializar el nodo receptor, calcular su clave pública y asimismo los parámetros de su clave privada. Ademas también se incluye la función de desencriptación del mensaje.

In [0]:
# Clase receptor 
class Receptor:

  def __init__(self, p, q):
    self.init_receptor(p, q )

  #Algoritmo de Euclides para el mcd
  def get_gcd(self, a, b):
    res_0  = a % b 
    if(res_0 == 0): return 0
    res_1 =  b % res_0 
    if(res_1 ==0 ): return 0
    res_2 = res_0 % res_1 
    temp_1 = res_1
    temp_2 = res_2
    temp=-1
    while(temp_2>0):
      temp = temp_1 % temp_2  
      temp_1= temp_2 
      temp_2= temp 
    return (temp_1)

  # Funcion que obtiene la llave d 
  def get_d_key(self, a, n ):
     c = 1
     while (c % a > 0):
      c += n
     return c // a

  # Inicializa el receptor, calculando la clave publica y la llave d.
  def init_receptor(self, p, q):
    self.n = p*q
    self.phi_n= (p-1)*(q-1)
    e = 0
    for i in range (15, self.phi_n):
      e= self.get_gcd(self.phi_n,i)
      if( e ==1 ):
          self.e= i
          break
    self.d = self.get_d_key(self.e, self.phi_n)

  # Funcion de desencriptacion 
  def decrypt(self, C):
    return pow(C, self.d, self.n) #Usa exponenciacion cuadratica modular


Luego, vamos a implementar la clase `Emisor` que permite configurar el emisor y encriptar el mensaje que quiere enviar.


In [0]:
class Emisor:
  def __init__ (self, e, n):
    self.e =e
    self.n =n 
  def encrypt(self, P):
    return pow(P,self.e,self.n) #Usa exponenciacion cuadratica modular

## Ejemplos

---

En los siguientes ejemplos se considera el [Codigo ASCII ](https://elcodigoascii.com.ar/) como equivalente numerico de los mensajes. 

En nuestro primer ejemplo, encriptaremos el siguiente mensaje: 
<br/>
*MATEMATICAS DISCRETAS*
<br/>
Asi, su equivalente en ASCII (en grupos de dos letras) sería

```
7765 8469 7765 8473 6765 8368 7383 6782 6984 6583
```


###**Ejemplo 1**



Inicializamos el receptor llamado `receptor` el cual comparte la informacion de clave pública, con valores de $p= 89 $, $q= 97 $, $n= 8633$

In [0]:
receptor = Receptor(89, 97) #Inicializamos un receptor R que comparte su clave publica
print("La clave pública del receptor R es: ({},{})".format(receptor.e, receptor.n))

La clave pública del receptor R es: (17,8633)


Compartimos la clave pública del receptor al emisor `emisor`, con la cual puede cifrar el mensaje que quiere enviar.

In [0]:
emisor = Emisor(receptor.e, receptor.n)

Con lo anterior, le pedimos al emisor `emisor` que cifre el mensaje, y que nos devuelva el mensaje cifrado (que posteriormente enviaremos a R)

In [0]:
message_to_encrypt = [7765, 8469, 7765, 8463, 6765, 8368, 7383, 6782, 6984, 6583]  #Obtenemos la informacion que envia el emisor
encrypted_message=[]  
for message in message_to_encrypt:
  encrypted_message.append(emisor.encrypt(message))
print("Este es el mensaje encriptado: ")
print(encrypted_message)

Este es el mensaje encriptado: 
[7164, 3074, 7164, 2531, 5430, 4425, 7563, 2126, 3492, 7292]


Posteriormente enviamos el mensaje al receptor R para que lo desencripte usando su clave privada. 
<br>
Al final, vemos que el mensaje es efectivamente el que hemos enviado: *Matemáticas Discretas*

In [0]:
decrypted_message= []
for message_encrypted in encrypted_message:
  decrypted_message.append( receptor.decrypt(message_encrypted))
print("Este es el mensaje desencriptado: ")
print(decrypted_message)

Este es el mensaje desencriptado: 
[7765, 8469, 7765, 8463, 6765, 8368, 7383, 6782, 6984, 6583]


###**Ejemplo 2**

A partir de este punto, generaremos primos aleatorios con el test de primalidad que vimos en la seccion: *Previo al cifrado*. <br/>

Le pediremos al algoritmo que genere numeros primos aleatorios más grandes, e iremos incrementando las cifras en los siguientes ejemplos. 


Para este ejemplo queremos encriptar el siguiente mensaje (con las reglas ya vistas): *ALGORITMO RSA*

    6576 7179 8273 8477 7982 8365

In [0]:
# Generamos numeros aleatorios de 4 cifras
p = get_prime(1000)
q = get_prime(1234)

Inicializamos el receptor llamado `receptor` el cual comparte la informacion de clave pública, con valores de $p$ y $q $ generados.

In [0]:
receptor = Receptor(p, q) #Inicializamos un receptor R que comparte su clave publica
print("La clave pública del receptor R es: ({},{})".format(receptor.e, receptor.n))

La clave pública del receptor R es: (17,3447163)


Compartimos la clave pública del receptor al emisor `emisor`, con la cual puede cifrar el mensaje que quiere enviar.

In [0]:
emisor = Emisor(receptor.e, receptor.n)

Con lo anterior, le pedimos al emisor `emisor` que cifre el mensaje, y que nos devuelva el mensaje cifrado (que posteriormente enviaremos a R)

In [0]:
message_to_encrypt = [7765, 8469, 7765, 8463, 6765, 8368, 7383, 6782, 6984, 6583]  #Obtenemos la informacion que envia el emisor
encrypted_message=[]  
for message in message_to_encrypt:
  encrypted_message.append(emisor.encrypt(message))
print("Este es el mensaje encriptado: ")
print(encrypted_message)

Este es el mensaje encriptado: 
[1295785, 1218521, 1295785, 1493066, 1706500, 1710384, 2131216, 427264, 2136521, 1359116]


Posteriormente enviamos el mensaje al receptor R para que lo desencripte usando su clave privada. 
<br>
Al final, vemos que el mensaje es efectivamente el que hemos enviado: *Matemáticas Discretas*

In [0]:
decrypted_message= []
for message_encrypted in encrypted_message:
  decrypted_message.append( receptor.decrypt(message_encrypted))
print("Este es el mensaje desencriptado: ")
print(decrypted_message)

Este es el mensaje desencriptado: 
[7765, 8469, 7765, 8463, 6765, 8368, 7383, 6782, 6984, 6583]


###**Ejemplo 3**

Para este ejemplo queremos encriptar el siguiente mensaje (con las reglas ya vistas): 
> *Un hombre provisto de papel, lápiz y goma, y con sujeción a una disciplina estricta, es en efecto una máquina de Turing universal*

Una gran frase de Alan Turing.

    85110 32104 111109 98114 10132 112114 111118 105115 116111 32100 10132 11297 112101 10844 32108 195161 112105 12232 12132 103111 10997 4432 12132 99111 11032 115117 106101 99105 195179 11032 9732 117110 9732 100105 11599 105112 108105 11097 32101 115116 114105 99116 9744 32101 11532 101110 32101 102101 99116 11132 117110 9732 109195 161113 117105 11097 32100 10132 84117 114105 110103 32117 110105 118101 114115 97108


In [0]:
# Generamos numeros aleatorios de 9 cifras
p = get_prime(100000000)
q = get_prime(123456789)

Inicializamos el receptor llamado `receptor` el cual comparte la informacion de clave pública, con valores de $p$ y $q $ generados.

In [0]:
receptor = Receptor(p, q) #Inicializamos un receptor R que comparte su clave publica
print("La clave pública del receptor R es: ({},{})".format(receptor.e, receptor.n))

La clave pública del receptor R es: (23,26821498133712209)


Compartimos la clave pública del receptor al emisor `emisor`, con la cual puede cifrar el mensaje que quiere enviar.

In [0]:
emisor = Emisor(receptor.e, receptor.n)

Con lo anterior, le pedimos al emisor `emisor` que cifre el mensaje, y que nos devuelva el mensaje cifrado (que posteriormente enviaremos a R)

In [0]:
message_to_encrypt = [85110, 32104, 111109, 98114, 10132, 112114, 111118, 
                      105115, 116111, 32100, 10132, 11297, 112101, 10844, 32108,
                      195161, 112105, 12232, 12132, 103111, 10997, 4432, 12132,
                      99111, 11032, 115117, 106101, 99105, 195179, 11032, 9732,
                      117110, 9732, 100105, 11599, 105112, 108105, 11097, 32101,
                      115116, 114105, 99116, 9744, 32101, 11532, 101110, 32101,
                      102101, 99116, 11132, 117110, 9732, 109195, 161113, 117105,
                      11097, 32100, 10132, 84117, 114105, 110103, 32117, 110105,
                      118101, 114115, 97108]
#Obtenemos la informacion que envia el emisor
encrypted_message=[]  
for message in message_to_encrypt:
  encrypted_message.append(emisor.encrypt(message))
print("Este es el mensaje encriptado: ")
print(encrypted_message)

Este es el mensaje encriptado: 
[5532485647077602, 24785113785478673, 20878629178507325, 20318147434924206, 17339841431785473, 4330874698266666, 1018515544965666, 4215149635318306, 1039598391495854, 19006712452659107, 17339841431785473, 20004841434577774, 860887740935743, 10324420587000223, 13051620192301000, 23945539579797579, 15093003054971839, 20854687271130672, 9229136722051410, 18138610040498656, 19433680682581510, 26160745253237248, 9229136722051410, 7862033585193373, 7890901913451938, 18155978251595174, 25730921063403841, 19261845409534157, 21831647416566879, 7890901913451938, 7529636785569122, 12728691359331635, 7529636785569122, 4548438873235465, 6212303089631946, 2955945919149282, 878118328006778, 6076208409941358, 23815526922121365, 4557260968729834, 16140707440925439, 24458613877332065, 14673265868958364, 23815526922121365, 20827199361078641, 8972139702122907, 23815526922121365, 12008277107761516, 24458613877332065, 4812685919445064, 12728691359331635, 7529636785569122, 432

Posteriormente enviamos el mensaje al receptor R para que lo desencripte usando su clave privada. 
<br>
Al final, vemos que el mensaje es efectivamente el que hemos enviado: *Matemáticas Discretas*

In [0]:
decrypted_message= []
for message_encrypted in encrypted_message:
  decrypted_message.append( receptor.decrypt(message_encrypted))
print("Este es el mensaje desencriptado: ")
print(decrypted_message)

Este es el mensaje desencriptado: 
[85110, 32104, 111109, 98114, 10132, 112114, 111118, 105115, 116111, 32100, 10132, 11297, 112101, 10844, 32108, 195161, 112105, 12232, 12132, 103111, 10997, 4432, 12132, 99111, 11032, 115117, 106101, 99105, 195179, 11032, 9732, 117110, 9732, 100105, 11599, 105112, 108105, 11097, 32101, 115116, 114105, 99116, 9744, 32101, 11532, 101110, 32101, 102101, 99116, 11132, 117110, 9732, 109195, 161113, 117105, 11097, 32100, 10132, 84117, 114105, 110103, 32117, 110105, 118101, 114115, 97108]
